File: cs2dart_roslyn_worker/Models/NormalizedSyntaxTree.cs
namespace Cs2DartRoslynWorker.Models;

/// <summary>
/// A rewritten syntax tree (plain-data nodes, no Roslyn types) paired with a SymbolTable.
/// </summary>
public sealed class NormalizedSyntaxTree
{
    /// <summary>Absolute path to the source file this tree was produced from.</summary>
    public string FilePath { get; set; } = string.Empty;

    /// <summary>
    /// The root node of the rewritten, annotated syntax tree.
    /// Represented as a generic object for the stub; full node types are defined in
    /// subsequent normalization pipeline tasks.
    /// </summary>
    public object? Root { get; set; }

    /// <summary>
    /// Maps every named-reference node in Root to its resolved symbol.
    /// Key: node identity (stable integer ID assigned during normalization).
    /// </summary>
    public SymbolTable SymbolTable { get; set; } = new();
}

File: lib/src/roslyn_frontend/models/normalized_syntax_tree.dart
import 'symbol_table.dart';
import 'syntax_node.dart';

/// A rewritten syntax tree (plain-data nodes, no Roslyn types) paired with
/// a [SymbolTable].
final class NormalizedSyntaxTree {
  /// Absolute path to the source file this tree was produced from.
  final String filePath;

  /// The root node of the rewritten, annotated syntax tree.
  ///
  /// Contains only plain-data node types — no Roslyn types.
  final SyntaxNode root;

  /// Maps every named-reference node in [root] to its resolved symbol.
  ///
  /// Key: node identity (stable integer ID assigned during normalization).
  /// Value: [ResolvedSymbol] record.
  final SymbolTable symbolTable;

  const NormalizedSyntaxTree({
    required this.filePath,
    required this.root,
    required this.symbolTable,
  });
}
